### PR TITLE
refactor: remove arguments from dialog fragment constructors

### LINKED ISSUE #4434: crashes when toggling light/dark mode while some menus are open
This form is for crashes when the comment section is open, more scenarios when crashes occur are in the replies below.

### Steps to reproduce

1. Set app theme to system
2. Go to a video that has comments
3. Open the comment section
4. Toggle system dark mode on or off (like by toggling battery saver or toggling dark mode in quick settings)

### Expected behavior

Switches between dark mode and light mode in the app, ideally without closing the comments section as that would be inconvenient

### Actual behavior

Crashes

### LibreTube version

0.16.1

### Android version

Android 13

### Other details
here is the log

```
java.lang.NullPointerException
	at com.github.libretube.ui.fragments.CommentsMainFragment.onViewCreated(SourceFile:77)
	at androidx.fragment.app.FragmentStateManager.createView(SourceFile:296)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(SourceFile:144)
	at androidx.fragment.app.FragmentManager.moveToState(SourceFile:62)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(SourceFile:36)
	at androidx.fragment.app.FragmentStateManager.createView(SourceFile:301)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(SourceFile:144)
	at androidx.fragment.app.FragmentManager.moveToState(SourceFile:62)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(SourceFile:36)
	at androidx.fragment.app.FragmentStateManager.createView(SourceFile:301)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(SourceFile:144)
	at androidx.fragment.app.FragmentManager.moveToState(SourceFile:62)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(SourceFile:36)
	at androidx.fragment.app.FragmentActivity.onStart(SourceFile:35)
	at androidx.appcompat.app.AppCompatActivity.onStart(SourceFile:1)
	at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1543)
	at android.app.Activity.performStart(Activity.java:8367)
	at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3669)
	at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:224)
	at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:204)
	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2306)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7918)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:952)
    
```

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by anshtya**
Addresses #4434

### CODE REVIEW

**Comment by Bnyro** [comment]
The same would probably also need to be done for everything in `.ui.sheets`, i.e. all bottom sheets since they're dialogs as well.

File: app/src/main/java/com/github/libretube/ui/dialogs/AddToPlaylistDialog.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/dialogs/AddToPlaylistDialog.kt:36]:
```suggestion
            videoId = it.getString("videoId")
```

File: app/src/main/java/com/github/libretube/ui/dialogs/DownloadDialog.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/dialogs/DownloadDialog.kt:41]:
```suggestion
        videoId = arguments.getString(IntentData.videoId)
```

File: app/src/main/java/com/github/libretube/ui/dialogs/PlaylistDescriptionDialog.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/dialogs/PlaylistDescriptionDialog.kt:30]:
```suggestion
            playlistId = it.getString(IntentData.playlistId)!!
```

**Comment by Bnyro** [comment]
Please save all the keys used for the bundles inside the `IntentData` object as a constant so that we can avoid typos in the key and keep the code cleaner.

**Comment by anshtya** [comment]
> Please save all the keys used for the bundles inside the `IntentData` object as a constant so that we can avoid typos in the key and keep the code cleaner.

Ok. Should I commit from here directly by clicking on 'Commit Suggestion' button or by pushing the changes to the branch?

**Comment by Bnyro** [comment]
It would be better if you do the suggested changes on your PC and push it via git as usual.

**Comment by anshtya** [comment]
> It would be better if you do the suggested changes on your PC and push it via git as usual.

Ok.

File: app/src/main/java/com/github/libretube/ui/dialogs/DownloadDialog.kt
Comment (anshtya) [reply app/src/main/java/com/github/libretube/ui/dialogs/DownloadDialog.kt:41]: It must be like this: `videoId = arguments?.getString(IntentData.videoId)!!`

Comment (Bnyro) [reply app/src/main/java/com/github/libretube/ui/dialogs/DownloadDialog.kt:41]: True

File: app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:128]:
```suggestion
                        val newDescription = resultBundle.getString(IntentData.playlistDescription)!!
```

File: app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:125]: Please save that to a constant somewhere (it'd be okay for me to add it to `IntentData` too.

File: app/src/main/java/com/github/libretube/constants/IntentData.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/constants/IntentData.kt:19]:
```suggestion
    const val playlistDescription = "playlistDescription"
```

**Comment by Bnyro** [comment]
Thanks for your work on that and applying the suggestions, looks fine to merge from my side when we can get the `FragmentResultListener` to work.

**Comment by Bnyro** [comment]
If you want to have a look at the bottom sheets as well: we sometimes pass the player there, the easiest thing probably would be to save the player in the `PlayerViewModel` and use it from there.

File: app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt
Comment (anshtya) [reply app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:125]: The `requestkey` ?

Comment (Bnyro) [reply app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:125]: Exactly

File: app/src/main/java/com/github/libretube/ui/dialogs/UpdateAvailableDialog.kt
Comment (Isira-Seneviratne) [inline app/src/main/java/com/github/libretube/ui/dialogs/UpdateAvailableDialog.kt:22]: It would be better to update the `UpdateInfo` class to implement the `Parcelable` interface instead (a custom `Parceler` implementation would be needed for `Instant`: https://developer.android.com/kotlin/parcelize#custom_parcelers). That way, this block could be replaced with the following:

```suggestion
        updateInfo = requireArguments().parcelable(IntentData.updateInfo)!!
```

File: app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt
Comment (Isira-Seneviratne) [inline app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt:108]: See the earlier comment.

```suggestion
                    val newUpdateAvailableDialog = UpdateAvailableDialog()
                    newUpdateAvailableDialog.arguments =
                        bundleOf(IntentData.updateInfo to updateInfo)
```

**Comment by anshtya** [comment]
> If you want to have a look at the bottom sheets as well: we sometimes pass the player there, the easiest thing probably would be to save the player in the `PlayerViewModel` and use it from there.

Ok.

File: app/src/main/java/com/github/libretube/ui/dialogs/ShareDialog.kt
Comment (anshtya) [reply app/src/main/java/com/github/libretube/ui/dialogs/ShareDialog.kt:39]: Please explain the reason for this suggestion.

Comment (Isira-Seneviratne) [reply app/src/main/java/com/github/libretube/ui/dialogs/ShareDialog.kt:39]: There are a couple of extensions that already do the API level check.

File: app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:124]: I think we should create the listener before we show the dialog?

Comment (anshtya) [reply app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:124]: It partially works. new description gets displayed only when the change description dialog is launched again by user.

**Comment by anshtya** [comment]
A couple of things I want to tell:
1. I have removed the arguments from the remaining dialog's constrcutors except the dialogs involving `PlaylistOptionsBottomSheet` due to some problem with `FragmentResultListener` and the `ColorPickerDialog`. I'll push the changes later.

2. In `ColorPickerDialog` there an argument for `Context`. I am not sure how to pass it through Bundle. I will think of its workaround later.

3. There is conflict in `app/src/main/java/com/github/libretube/ui/preferences/MainSettings.kt` due to the recent push in `master`. I am not sure what to do for it. Resolve Conflicts button is appearing for me in this PR.

**Comment by Bnyro** [comment]
1. Sounds good :+1:
2. We can't pass the context like that, however bottom sheets should have its own context after `onCreate` was called. So just make everything that is initialized with the context `lateinit var` and create it in the `onCreate` function.
3. No worries about that, I can take care of it when merging.

**Comment by anshtya** [comment]
Why status checks aren't running ?

**Comment by Bnyro** [comment]
Probably due to the merge conflicts.
I don't have my PC with me currently, but in theory you can just click the resolve conflicts button and remove the two or three lines indicating that there's a conflict (it should be fine to just keep all the code as it is there).

File: app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt
Comment (anshtya) [reply app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:124]: I got it working. I attached the fragment listener in the adapter, in this case the `PlaylistAdapter`.
 
`holder.binding.apply { `
` //some other code`
`val fragmentManager = (root.context as BaseActivity).supportFragmentManager`
` fragmentManager.setFragmentResultListener(`
`                IntentData.requestKey,`
`               (root.context as BaseActivity) `-> is this correct LifecycleOwner for it according to you?
`           ) { _, resultBundle ->`

Comment (Bnyro) [reply app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:124]: Seems fine if it works :)

Comment (anshtya) [reply app/src/main/java/com/github/libretube/ui/sheets/PlaylistOptionsBottomSheet.kt:124]: Yes it works.

**Comment by anshtya** [comment]
> 1. Sounds good 👍
> 2. We can't pass the context like that, however bottom sheets should have its own context after `onCreate` was called. So just make everything that is initialized with the context `lateinit var` and create it in the `onCreate` function.
> 3. No worries about that, I can take care of it when merging.

2. `ColorPickerDialog` is working as expected without supplying the context. It only uses to context for showing toast messages.
3.  Could you please check the `MainSettings.kt` that I have pushed?

Everything is done. If there aren't any things need to be corrected from your side, I will push the remaining Dialogs.

**Comment by Bnyro** [comment]
`MainSettings.kt` looks good to me :+1:

**Comment by anshtya** [comment]
Is there something wrong with `IntentData.kt`, code quality check is failing.

**Comment by Bnyro** [comment]
There's a package called `ktlint` for Linux, with that one installed you can run `ktlint --android -F` to automatically format the code.

**Comment by anshtya** [comment]
> There's a package called `ktlint` for Linux, with that one installed you can run `ktlint --code-style=android-studio -F` to automatically format the code.

Ktlint doesn't show anything wrong in IntentData for me. I use windows.

**Comment by anshtya** [comment]
> There's a package called `ktlint` for Linux, with that one installed you can run `ktlint --code-style=android-studio -F` to automatically format the code.

I have tried placing the `IntentData` in a new Android Project with ktlint setup and it. After doing the ktlint format command the code in `IntentData` remains same and doesn't give any errors.

File: app/src/main/java/com/github/libretube/constants/IntentData.kt
Comment (Bnyro) [inline app/src/main/java/com/github/libretube/constants/IntentData.kt:6]:
```suggestion
    const val ID = "id"
```

**Comment by Bnyro** [comment]
It asks us to use variable names like `PLAYLIST_DESCRIPTION` instead of `playlistDescription`, but we can safely ignore this.

**Comment by anshtya** [comment]
> It asks us to use variable names like `PLAYLIST_DESCRIPTION` instead of `playlistDescription`, but we can safely ignore this.

Yes. So what should I do next?

**Comment by Bnyro** [comment]
If everything works from your perspective, we can unmark it as draft.
However, I'd like to test this a bit myself (and others as well) first before merging to avoid any new issues by the changes.

**Comment by anshtya** [comment]
> If everything works from your perspective, we can unmark it as draft. However, I'd like to test this a bit myself (and others as well) first before merging to avoid any new issues by the changes.

Sure. Just send me a reply for any changes or to unmark from draft after you have tested.

**Comment by Bnyro** [comment]
1. When renaming a playlist or changing its description it's not visible in the UI immediately as it used to, the playlist has to be reloaded which is not ideal.
2. When changing the theme mode, all crashes while a dialog is opened seem to be gone (which is great). However, we still need to do the same thing for all bottom sheets to fix the issue, but that can be done by someone else in a new PR.

**Comment by anshtya** [comment]
> 1. When renaming a playlist or changing its description it's not visible in the UI immediately as it used to, the playlist has to be reloaded which we should fix before merging.
> 2. When changing the theme mode, all crashes while a dialog is opened seem to be gone (which is great). However, we still need to do the same thing for all bottom sheets to fix the issue, but that can be done by someone else in a new PR.
> 
> If 1. is fixed, it's okay to be merged from my POV.

1.  Please recheck. It is updated immediately.

**Comment by Bnyro** [comment]
No, that's clearly not working as it does before the changes.

**Comment by Bnyro** [comment]
We can consider passing the callback functions differently instead of using the `FragmentResultListener`:
```kotlin
val renamePlaylistDialog = RenamePlaylistDialog()
renamePlaylistDialog.onRename = {
    // insert the callback function content here
}
renamePlaylistDialog.show(fragmentManager, "foo")
```
where `onRename` is a public variable of type `() -> Unit`.

**Comment by anshtya** [comment]
> We can consider passing the callback functions differently instead of using the `FragmentResultListener`:
> 
> ```kotlin
> val renamePlaylistDialog = RenamePlaylistDialog()
> renamePlaylistDialog.onRename = {
>     // insert the callback function content here
> }
> renamePlaylistDialog.show(fragmentManager, "foo")
> ```
> 
> where `onRename` is a public variable of type `() -> Unit`.

Ok.

**Comment by anshtya** [comment]
> We can consider passing the callback functions differently instead of using the `FragmentResultListener`:
> 
> ```kotlin
> val renamePlaylistDialog = RenamePlaylistDialog()
> renamePlaylistDialog.onRename = {
>     // insert the callback function content here
> }
> renamePlaylistDialog.show(fragmentManager, "foo")
> ```
> 
> where `onRename` is a public variable of type `() -> Unit`.

Done.

**Comment by anshtya** [comment]
> 1. When renaming a playlist or changing its description it's not visible in the UI immediately as it used to, the playlist has to be reloaded which we should fix before merging.
> 2. When changing the theme mode, all crashes while a dialog is opened seem to be gone (which is great). However, we still need to do the same thing for all bottom sheets to fix the issue, but that can be done by someone else in a new PR.
> 
> If 1. is fixed, it's okay to be merged from my POV.

Any other changes ?

**Review by Bnyro — APPROVED** [review]
Thank you!

**Comment by anshtya** [comment]
> Are you interested in looking into the bottom sheets too, or shall I do that?

I would but I am kinda busy nowadays due to college. Thanks.

**Comment by Bnyro** [comment]
We're not in a hurry, you can do it whenever you want :)

**Comment by anshtya** [comment]
> We're not in a hurry, you can do it whenever you want :)

Ok. Sounds good.

**Comment by Bnyro** [comment]
For your information, I've already done the changes in some separate PRs now.

Nevertheless, feel free to work on any other issue whenever you want to :)

**Comment by anshtya** [comment]
> For your information, I've already done the changes in some separate PRs now.
> 
> Nevertheless, feel free to work on any other issue whenever you want to :)

Ok 👍

**Pull request merged**
